clients/venice-client/src/main/java/com/linkedin/venice/fastclient/transport/GrpcTransportClient.java

```diff
@@ -0,0 +1,182 @@
+import com.linkedin.venice.client.store.transport.TransportClientResponse;
+import com.linkedin.venice.compression.CompressionStrategy;
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.fastclient.ClientConfig;
+import com.linkedin.venice.protocols.VeniceClientRequest;
+import com.linkedin.venice.protocols.VeniceReadServiceGrpc;
+import com.linkedin.venice.protocols.VeniceServerResponse;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class GrpcTransportClient extends InternalTransportClient {
+  private static final Logger LOGGER = LogManager.getLogger(GrpcTransportClient.class);
+  private static final String STORAGE_ACTION = "storage";
+  private final VeniceConcurrentHashMap<String, ManagedChannel> serverGrpcChannels;
+  private final Map<String, String> nettyAddressToGrpcAddressMap;
+  // we cache stubs to avoid creating a new stub for each request, improves performance
+  private final VeniceConcurrentHashMap<ManagedChannel, VeniceReadServiceGrpc.VeniceReadServiceStub> stubCache;
+  private final TransportClient r2TransportClient; // used for non-storage related requests
+
+  public GrpcTransportClient(ClientConfig clientConfig) {
+    serverGrpcChannels = new VeniceConcurrentHashMap<>();
+    stubCache = new VeniceConcurrentHashMap<>();
+    nettyAddressToGrpcAddressMap = clientConfig.getNettyServerToGrpcAddressMap();
+    this.r2TransportClient = new R2TransportClient(clientConfig.getR2Client());
+  }
+
+  protected ManagedChannel getChannel(String serverAddress) {
+    if (!nettyAddressToGrpcAddressMap.containsKey(serverAddress)) {
+      throw new VeniceException("No grpc server found for address: " + serverAddress);
+    }
+
+    return serverGrpcChannels.computeIfAbsent(
+        serverAddress,
+        k -> ManagedChannelBuilder.forTarget(nettyAddressToGrpcAddressMap.get(k)).usePlaintext().build());
+  }
+
+  protected VeniceReadServiceGrpc.VeniceReadServiceStub getStub(ManagedChannel channel) {
+    return stubCache.computeIfAbsent(channel, VeniceReadServiceGrpc::newStub);
+  }
+
+  @Override
+  public CompletableFuture<TransportClientResponse> get(String requestPath, Map<String, String> headers) {
+    String[] requestParts = requestPath.split("/");
+    if (!requestParts[3].equals(STORAGE_ACTION)) {
+      LOGGER.debug(
+          "performing unsupported gRPC transport client action ({}), passing request to R2 client",
+          requestParts[3]);
+      return r2TransportClient.get(requestPath, headers);
+    }
+
+    ManagedChannel channel = getChannel(requestParts[2]);
+
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
+        .setResourceName(requestParts[4])
```

Comment:
Does GRPC have capability to carry request headers to the backend?
Such as client compression or compute header, and it will be good to be able to carry these headers to the backend.

Reply:
We can pass headers in this manner via attaching metadata to the response with a gRPC service interceptor. I will look into this.

Reply:
We can also consider adding it as part of the API itself

Reply:
I will address in a later PR.